physicsobjects.hxx

```diff
@@ -203,6 +203,57 @@ auto ObjectMassCorrectionWithPt(auto df, const std::string corrected_mass,
     return df1;
 }
 
+/// Function to correct object mass in alignment with object pt correction
+///
+/// \param[in] df the input dataframe
+/// \param[out] output_flag the name of the corrected masses to be determined
+/// \param[in] leptons_pt name of the input pt column of the lepton collection
+/// \param[in] leptons_eta name of the input eta column of the lepton collection
+/// \param[in] leptons_phi name of the input phi column of the lepton collection
+/// \param[in] leptons_mass name of the input mass column of the lepton
+/// collection \param[in] leptons_charge name of the input charge column of the
+/// lepton collection \param[in] leptons_mask name of the input mask column of
+/// the lepton collection that marks lepton to be taken into account \param[in]
+/// dR_cut minimum required angular distance between the leptons
+///
+/// \return a dataframe containing the modified object masses
+auto CheckForDiLeptonPairs(auto df, const std::string output_flag,
+                           const std::string leptons_pt,
+                           const std::string leptons_eta,
+                           const std::string leptons_phi,
+                           const std::string leptons_mass,
+                           const std::string leptons_charge,
+                           const std::string leptons_mask, const float dR_cut) {
+    auto pair_finder_lambda = [dR_cut](const ROOT::RVec<float> &pt_values,
+                                       const ROOT::RVec<float> &eta_values,
+                                       const ROOT::RVec<float> &phi_values,
+                                       const ROOT::RVec<float> &mass_values,
+                                       const ROOT::RVec<int> &charge_values,
+                                       const ROOT::RVec<int> &mask) {
+        const auto valid_lepton_indices = ROOT::VecOps::Nonzero(mask);
+        for (auto it1 = valid_lepton_indices.begin();
+             it1 != valid_lepton_indices.end(); it1++) {
+            for (auto it2 = it1 + 1; it2 != valid_lepton_indices.end(); it2++) {
+                if (charge_values.at(*it1) != charge_values.at(*it2)) {
+                    auto p4_1 = ROOT::Math::PtEtaPhiMVector(
+                        pt_values.at(*it1), eta_values.at(*it1),
+                        phi_values.at(*it1), mass_values.at(*it1));
+                    auto p4_2 = ROOT::Math::PtEtaPhiMVector(
+                        pt_values.at(*it2), eta_values.at(*it2),
+                        phi_values.at(*it2), mass_values.at(*it2));
+                    if (ROOT::Math::VectorUtil::DeltaR(p4_1, p4_2) >= dR_cut)
```

Comment:
I was wondering, is this faster or slower compared to a version, where you calculate the deltaR for every iteration and store the bool if the deltaR requirement is passed and then return an any of this vector, so basically running without the if statement here

Reply:
I think in this case the clear advantage is that the loop stops as soon as one suitable pair is found.

Reply:
Ok, most events don't have such a pair and the full loop runs. But I think that the Any function must do something similar internally. If you compare directly, you save the memory access for the result of the comparison.